rec/utils.py

```diff
@@ -9,13 +9,21 @@
 
 
 def normalize_matrix(matrix, axis=1):
-    divisor = matrix.sum(axis=axis)[:, None]
-    result = np.divide(
-        matrix, divisor, out=np.zeros(matrix.shape, dtype=float), where=divisor != 0
-    )
+    """ Normalize a matrix so that each row vector has a Euclidean norm of 1.
+    """
+    divisor = np.linalg.norm(matrix, axis=1)[:, np.newaxis]
+    # only normalize where divisor is not zero
+    result = np.divide(matrix, divisor, out=np.zeros(matrix.shape), where=divisor != 0)
     return result
 
 
+def contains_row(matrix, row):
+    """ Check if a numpy matrix contains a row with the same values as the 
+        variable `row`.
+    """
+    return (matrix == row).all(1).any()
```

Comment:
Is `all(1)` equivalent to `all(axis=1)`?
Why do we need both `all` and `any`?

Reply:
yes! the notation is equivalent, but I'll add the `axis` keyword argument so it's more clear. We need both `all` and `any` because `all` is checking each row for equality while the subsequent `any` checks if at least one row matches. maybe this will clear it up:

![image](https://user-images.githubusercontent.com/5581093/93247344-07907a80-f743-11ea-9ec9-a97a25c7ddb3.png)

Reply:
Gotcha, thanks!